engine/src/core/platform/graphics/rendering/rendering_device/buffers/opengl_vertex_array.cpp

```diff
@@ -0,0 +1,49 @@
+//
+// Created by blinku on 22.07.19.
+//
+
+#include <algorithm>
+#include "opengl_vertex_array.h"
+#include "opengl_buffer.h"
+
+#include "glad/glad.h"
+
+namespace Bald::Platform::Graphics {
+
+    OpenGLVertexArray::OpenGLVertexArray() noexcept : m_ArrayID(0) {
+        glGenVertexArrays(1, &m_ArrayID);
+    }
+
+    OpenGLVertexArray::~OpenGLVertexArray() {
+
+        std::for_each(m_Buffers.begin(), m_Buffers.end(), [](Bald::Graphics::Buffer* buffer) { delete buffer; });
+
+        glDeleteVertexArrays(1, &m_ArrayID);
+    }
+
+    void OpenGLVertexArray::AddBuffer(Bald::Graphics::Buffer* buffer, [[maybe_unused]] unsigned index) noexcept {
+
+        Bind();
+        buffer->Bind();
+
+        glEnableVertexAttribArray(index);
+        glVertexAttribPointer(index, static_cast<int>(buffer->GetComponentCount()), GL_FLOAT, GL_FALSE, static_cast<int>(buffer->GetComponentCount() * sizeof(float)), nullptr);
+
+        m_Buffers.push_back(buffer);
+
+        // As far as I remember we shouldn't unbind the buffer here. This will end up in it not being bound to vertex array object.
+        // Gotta check this.
+        // buffer->Unbind();
```

Comment:
We should.

Reply:
"A VAO stores the glBindBuffer calls when the target is GL_ELEMENT_ARRAY_BUFFER. This also means it stores its unbind calls so make sure you don't unbind the element array buffer before unbinding your VAO, otherwise it doesn't have an EBO configured."
https://learnopengl.com/Getting-started/Hello-Triangle

Reply:
so If I understand correctly you should do it.

Reply:
but you should do it after unbinding VAO